src/util/sharing_node.h

```diff
@@ -12,339 +12,455 @@ Author: Daniel Poetzl
 #ifndef CPROVER_UTIL_SHARING_NODE_H
 #define CPROVER_UTIL_SHARING_NODE_H
 
-#include <list>
+//#define DEBUG
+
+#ifdef DEBUG
+#include <iostream>
+#endif
+
+#include <forward_list>
+#include <type_traits>
+
+#ifndef SN_SMALL_MAP
+#define SN_SMALL_MAP 1
+#endif
+
+#ifndef SN_SHARE_KEYS
+#define SN_SHARE_KEYS 0
+#endif
+
+#if SN_SMALL_MAP == 1
+#include "small_map.h"
+#else
 #include <map>
-#include <memory>
+#endif
 
 #include "invariant.h"
+#include "make_unique.h"
+#include "small_shared_ptr.h"
+#include "small_shared_two_way_ptr.h"
+
+//#define SN_INTERNAL_CHECKS
```

Comment:
Part of the above comment also applies to this as well? Can this be documented somewhere that it can be passed during compilation with `-DSN_INTERNAL_CHECKS` and removed from here? Otherwise it's confusing to a new reader, I believe.